python/tk_houdini/bootstrap.py

```diff
@@ -12,68 +12,274 @@
+        import hou
+        hou.hipFile.load(file_to_open)
+
+    # clean env vars. note, we don't clean the temp dir env variable since it is
+    # used by the engine to know where to write the shelf/menu .xml files
+    for env_var in [g_sgtk_context_env, g_sgtk_engine_env, g_sgtk_file_to_open_env]:
+        if env_var in os.environ:
+            del os.environ[env_var]
+
+
+def bootstrap_exception(error_msg):
+    """
+    Shows an error message if there is a problem during bootstrap.
+    """
+
+    # this file can be imported before houdini is up and running, but this
+    # method should only be called during a houdini session
+    import hou
+
+    # get a full stack trace
+    details = traceback.format_exc()
+
+    if hou.isUIAvailable():
+        # we have a UI, show the error in a popup dialog
+        hou.ui.displayMessage(error_msg, details=details)
+    else:
+        # no UI, just print to stdout
+        print error_msg
+        print details
+
+
+################################################################################
+# utility methods for populating the environment prior to bootstrap
+
+def get_classic_startup_env():
+    """
+    Returns a dict of key/value pairs representing the environment variables
+    needed to launch houdini and startup toolkit in classic mode.
+    """
+
+    # Add the classic startup directory for the engine (2 levels up from this
+    # file). This directory, in the engine root, includes the pythonrc.py files
+    # within python2.6libs and python2.7libs directories once on HOUDINI_PATH,
+    # houdini will execute the appropriate file at startup
+    startup_path = os.path.join(
+        os.path.dirname(__file__),
+        "..",
+        "..",
+        "startup"
+    )
+    startup_path = os.path.normpath(startup_path)
+
+    # update the houdini path with the engine startup path
+    return _get_env([startup_path])
+
+
+def get_plugin_startup_env(plugin_names):
+    """
+    Returns a dict of key/value pairs representing the environment variables
+    needed to launch houdini with the supplied plugins.
+    """
+
+    # for each plugin provided, construct a list of corresponding plugin startup
+    # paths.
+    startup_paths = []
+
+    for plugin_name in plugin_names:
+
+        # Each plugin should have a startup directory at the top level. this
+        # directory should include the standard pythonX.Xlibs/pythonrc.py setup
+        # which houdini will execute at startup time.
+        plugin_startup_path = os.path.join(
+            os.path.dirname(__file__),
+            "..",
+            "..",
+            "plugins",
+            plugin_name,
+            "startup"
+        )
+        plugin_startup_path = os.path.normpath(plugin_startup_path)
+
+        # if the path exists, add it to the list of startup paths
+        if os.path.exists(plugin_startup_path):
+            startup_paths.append(plugin_startup_path)
+
+    return _get_env(startup_paths)
```

Comment:
The only real difference between classic and plugin startup is the list of paths added to `HOUDINI_PATH` for automatic launch at startup. In classic mode, the startup logic lives in `startup/python2.X/pythonrc.py` in the engine root. In the plugin case, there are similar files in each plugin root.